### ISSUE TITLE
pangolin --all-versions broken with 4.1

### ISSUE THREAD

**Issue by HadrienG** [cov-lineages/pangolin#467]
I'm running pangolin as part of wf-artic (with docker, image: `ontresearch/pangolin:4.0.5`). It failed on this morning's samples:

pangolin 4.0.5:

```bash
epi2melabs@dad1d3221ca3:~$ pangolin -v
pangolin 4.0.5
epi2melabs@dad1d3221ca3:~$ pangolin --all-versions
pangolin: 4.0.5
pangolin-data: 1.3
constellations: v0.1.6
scorpio: 0.3.16
```

after running pangolin --update:

```bash
epi2melabs@dad1d3221ca3:~$ pangolin --update
Latest for pangolin is v4.1
pangolin updated to v4.1
Latest for pangolin-data is v1.11
pangolin-data updated to v1.11
Latest for constellations is v0.1.10
constellations updated to v0.1.10
Latest for scorpio is v0.3.17
scorpio updated to v0.3.17
epi2melabs@dad1d3221ca3:~$ pangolin -v
pangolin 4.1
epi2melabs@dad1d3221ca3:~$ pangolin --all-versions
pangolin: 4.1
pangolin-data: 1.11
constellations: v0.1.10
scorpio: 0.3.17
Traceback (most recent call last):
  File "/home/epi2melabs/conda/bin/pangolin", line 8, in <module>
    sys.exit(main())
  File "/home/epi2melabs/conda/lib/python3.8/site-packages/pangolin/command.py", line 156, in main
    print_versions_exit(config)
  File "/home/epi2melabs/conda/lib/python3.8/site-packages/pangolin/utils/initialising.py", line 226, in print_versions_exit
    print_conda_version(['usher', 'ucsc-fatovcf', 'gofasta', 'minimap2'])
  File "/home/epi2melabs/conda/lib/python3.8/site-packages/pangolin/utils/initialising.py", line 203, in print_conda_version
    result = subprocess.run(['conda', 'list', pkg],
  File "/home/epi2melabs/conda/lib/python3.8/subprocess.py", line 493, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/home/epi2melabs/conda/lib/python3.8/subprocess.py", line 858, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/home/epi2melabs/conda/lib/python3.8/subprocess.py", line 1704, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'conda'
```

**Comment by aineniamh** [comment]
Thanks for flagging this @HadrienG and @kapsakcj. I think you might be right about the conda thing, I'll have a look now.

**Comment by kapsakcj** [comment]
> Although saying that, our github actions use mamba and seem to be working still.

If the GH actions environment includes `conda` alongside `mamba` (for example if `conda` is used to install `mamba`) then that may be why the GH actions workflow is running successfully. Hard to say though, I'm not too familiar with this GH actions workflow: https://github.com/cov-lineages/pangolin/blob/78c24e83b4b37476e8e6bc53e3a86ca3ff5dcb8a/.github/workflows/pangolin.yml#L16


The name of it makes me think that miniconda is first installed, which comes with `conda` executable

**Comment by fmaguire** [comment]
replacing https://github.com/cov-lineages/pangolin/blob/78c24e83b4b37476e8e6bc53e3a86ca3ff5dcb8a/pangolin/utils/initialising.py#L226 with

```
for program in ['usher --version', 'gofasta --version', 'minimap2 --version',]:
    output = subprocess.run(program, shell=True, check=True,
                             stdout=subprocess.PIPE, encoding='utf-8')
    version = output.stdout.strip().split()[-1].strip('()v')
    print(f"{program.split()[0]} {version}")
```

Would be a more robust/less conda dependent way to easily get the versions of everything apart from `faToVcf` (which doesn't have a version function unfortunately but may be worth asking ucsc about adding it to their fork).  This methods means the correct tool version (i.e., the one pangolin will actually try to use) will be returned in situations where someone has messed up their conda path/config!

**Comment by aineniamh** [comment]
Ah that makes sense so! I've got a potential fix here for your case @kapsakcj, but It assumes that people will either have conda or mamba installed though so not a fix for the original poster. 

Ah, @fmaguire just seeing your message there- yeah this is what we had before. Maybe we should just revert to this style.

**Comment by kapsakcj** [comment]
Thanks @fmaguire and @aineniamh . I know `usher --version` exists, I requested that flag a while back 😄 

gofasta also has that flag
```
$ gofasta --version
gofasta version 1.0.0
```
the tricky one might be `ucsc-fatovcf`

**Comment by kapsakcj** [comment]
That commit (aca98ff) fixed it for me

**Comment by aineniamh** [comment]
https://github.com/cov-lineages/pangolin/releases/tag/v4.1.1
This tag has the updated fix, but will update again when faToVcf can be included

**Comment by AngieHinrichs** [comment]
Reading the [code](https://github.com/ucscGenomeBrowser/kent/blob/5e7ae9c0bddd97efa236e2a138e00e9ccc5dda1e/src/lib/options.c#L329), the kent/src options processing lib module actually has a `--version`-ish behavior already, with a characteristically odd invocation:
```
faToVcf -verbose=2 -h
```
The first line of output includes the version:
```
### kent source version 426 ###
```

**Comment by aineniamh** [comment]
```
def print_faToVf_version():
    output = subprocess.run("faToVcf -verbose=2 -h 2>&1 | grep '#'", shell=True, check=True,
                                stdout=subprocess.PIPE, encoding='utf-8')
    version = output.stdout.split(' ')[-2]
    print(f"faToVcf: {version}")
```
I've added this in as a potential version print out? 
Running it gives
```
pangolin: 4.1
pangolin-data: 1.11
constellations: v0.1.10
scorpio: 0.3.17
usher 0.5.4
gofasta 1.1.0
minimap2 2.24-r1122
faToVcf: 426
```

**Comment by AngieHinrichs** [comment]
Excellent, thanks @aineniamh!

**Comment by kapsakcj** [comment]
Thanks for the quick fix @aineniamh and thank you for the explanation @AngieHinrichs ! Sorry @pvanheus , I didn't mean to point fingers. I clearly don't understand how to dig into python tracebacks

The 4.1.1 release has resolved this issue for me.

**Comment by aineniamh** [comment]
Latest release has this issue fixed in it now [v4.1.2](https://github.com/cov-lineages/pangolin/releases/tag/v4.1.2).
